### PR TITLE
python312Packages.nose: fix build

### PR DESCRIPTION
**Pull request by jchv**
## Description of changes

This is an alternative to #325935 which uses a patch that I made independently without looking at the other patches. If we can't figure that out, then this is a backup plan.

I anticipate that my patch is incomplete, but it is complete enough to pass checks on all of the packages that I just re-enabled checks for.

This also fixes Python 3.13 support in Nose. Nose relies on 2to3, which is removed in Python 3.13, so we always use 3to2 from Python 3.12. This should buy us some more time to wait for upstreams to either get off of Nose or get removed for being unmaintained.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/nose/0002-nose-python-3.12-fixes.patch
Comment (emilazy) [inline pkgs/development/python-modules/nose/0002-nose-python-3.12-fixes.patch:28]: This is just a nit because the practice is really common, but:

```
2. Subject to the terms and conditions of this License Agreement, PSF hereby
grants Licensee a nonexclusive, royalty-free, world-wide license to reproduce,
analyze, test, perform and/or display publicly, prepare derivative works,
distribute, and otherwise use Python alone or in any derivative version,
provided, however, that PSF's License Agreement and PSF's notice of copyright,
i.e., "Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023 Python Software Foundation;
All Rights Reserved" are retained in Python alone or in any derivative version
prepared by Licensee.

3. In the event Licensee prepares a derivative work that is based on
or incorporates Python or any part thereof, and wants to make
the derivative work available to others as provided herein, then
Licensee hereby agrees to include in any such work a brief summary of
the changes made to Python.
```

Would you be willing to make the patch add the full upstream `LICENSE` file as `LICENSE.cpython` file, and adapt the comment similarly to the following?

```
# Adapted from the CPython 3.11 imp.py code.
# Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023 Python Software Foundation; All Rights Reserved
# Originally licensed under the PSLv2 (see LICENSE.cpython) and incorporated under the LGPL 2.1 (see lgpl.txt).
```

File: pkgs/development/python-modules/nose/default.nix
Comment (emilazy) [inline pkgs/development/python-modules/nose/default.nix:37]: I guess when 3.12 leaves Nixpkgs we’ll have to do one of:

* package it independently of the Python interpreter;
* run it once over the `nose` code and just vendor the results; or
* remove `nose`?

Comment (mweinelt) [reply pkgs/development/python-modules/nose/default.nix:37]: Idealler by 2028 we've fixed the ecosystem enough, that nose is not relevant anymore.

Comment (emilazy) [reply pkgs/development/python-modules/nose/default.nix:37]: I didn’t realize that we kept Python releases for that many years. Agreed that it’s not worth thinking about in that case.

File: pkgs/development/python-modules/nose/0002-nose-python-3.12-fixes.patch
Comment (mdmintz) [reply pkgs/development/python-modules/nose/0002-nose-python-3.12-fixes.patch]: Hopefully all is set on the `pynose` side from https://github.com/mdmintz/pynose/issues/33#issuecomment-2221149163. Let me know if anything else!

**Comment by mweinelt** [comment]
Due to the rebuild count this should target the staging branch.

**Comment by emilazy** [comment]
I am also running a `nixpkgs-review` run on the community builder, but I don’t know if I’ll get better outcomes than you :)

I feel more confident in this approach than https://github.com/NixOS/nixpkgs/pull/325935. The other patch seems incomplete too and i still feel more assured about the provenance of this one.

**Comment by emilazy** [comment]
I guess we can use the time it’ll take this to get through staging to hunt down unnecessary uses of `nose` throughout the tree.

**Comment by emilazy** [comment]
You brought some `master` commits in, looks like it’ll need a rebase (or wait until the next `master` → `staging` merge?).

**Comment by mweinelt** [comment]
> That said it would probably be good to wait for master to be merged back into staging, since I can't re-enable the tests in packages if they haven't been disabled yet, but I can fix that later.

Done.

**Comment by emilazy** [comment]
I forgot that things like Tensorflow would be in the build chain…

It was at 78 failures after >1,400 successful builds and 239 builds left to finish, which seems good to me. I forgot to start the job in the builder in `tmux` so I’ll leave it at that for now. If you get a full `nixpkgs-review` report I can do an automated check against the latest Hydra unstable job and the one before the 3.12 merge.

**Comment by natsukium** [comment]
Since `nose` is a test suite and does not propagate, I guess it is not necessary to build many other things if those that directly depend on `nose` can be built.

**Review by mweinelt — APPROVED** [review]

**Comment by mweinelt** [comment]
This change has to pass through staging first. Check back in two weeks.

https://github.com/NixOS/rfcs/blob/master/rfcs/0026-staging-workflow.md

**Pull request merged**
